src/input.rs

```diff
@@ -0,0 +1,497 @@
+    map: HashMap<T, Button>,
+}
+
+impl<T> Mapper<T>
+where
+    T: Copy + Eq + std::hash::Hash
+{
+    pub fn new() -> Self {
+        Self {
+            map: HashMap::new(),
+        }
+    }
+
+    /// Add a new action to mapper. If action already exists, it will be overridden.
+    pub fn add_action(&mut self, action: T, button: Button) {
+        self.map.insert(action, button);
+    }
+
+    /// Add multiple actions to mapper. Existing actions will be overridden.
+    pub fn add_actions(&mut self, actions: Vec<(T, Button)>) {
+        for (action, button) in actions.iter() {
+            self.map.insert(*action, *button);
+        }
+    }
+
+    /// Get button that is binded to that action.
+    pub fn get_button(&self, action: T) -> Option<&Button> {
+        self.map.get(&action)
+    }
+
+    /// Remove action from mapper.
+    pub fn remove_action(&mut self, action: T) {
+        self.map.remove(&action);
+    }
+
+    /// Remove multiple actions from mapper.
+    pub fn remove_actions(&mut self, actions: Vec<T>) {
+        for action in actions.iter() {
+            self.map.remove(action);
+        }
+    }
+
+    /// Removes all actions and set new ones.
+    pub fn set(&mut self, actions: Vec<(T, Button)>) {
+        self.map.clear();
+        self.add_actions(actions);
+    }
+}
+
+/// Symbolic name for a keyboard key.
+#[derive(Debug, Hash, Ord, PartialOrd, PartialEq, Eq, Clone, Copy, TryFromPrimitive)]
+#[repr(u32)]
+pub enum KeyCode {
+    /// The '1' key over the letters.
+    Key1,
+    /// The '2' key over the letters.
+    Key2,
+    /// The '3' key over the letters.
+    Key3,
+    /// The '4' key over the letters.
+    Key4,
+    /// The '5' key over the letters.
+    Key5,
+    /// The '6' key over the letters.
+    Key6,
+    /// The '7' key over the letters.
+    Key7,
+    /// The '8' key over the letters.
+    Key8,
+    /// The '9' key over the letters.
+    Key9,
+    /// The '0' key over the 'O' and 'P' keys.
+    Key0,
+
+    A,
+    B,
+    C,
+    D,
+    E,
+    F,
+    G,
+    H,
+    I,
+    J,
+    K,
+    L,
+    M,
+    N,
+    O,
+    P,
+    Q,
+    R,
+    S,
+    T,
+    U,
+    V,
+    W,
+    X,
+    Y,
+    Z,
+
+    /// The Escape key, next to F1.
+    Escape,
+
+    F1,
+    F2,
+    F3,
+    F4,
+    F5,
+    F6,
+    F7,
+    F8,
+    F9,
+    F10,
+    F11,
+    F12,
+    F13,
+    F14,
+    F15,
+    F16,
+    F17,
+    F18,
+    F19,
+    F20,
+    F21,
+    F22,
+    F23,
+    F24,
+
+    /// Print Screen/SysRq.
+    PrintScreen,
+    /// Scroll Lock.
+    ScrollLock,
+    /// Pause/Break key, next to Scroll lock.
+    Pause,
+
+    /// `Insert`, next to Backspace.
+    Insert,
+    Home,
+    Delete,
+    End,
+    PageDown,
+    PageUp,
+
+    Left,
+    Up,
+    Right,
+    Down,
+
+    /// The Backspace key, right over Enter.
+    Backspace,
+    /// The Enter key.
+    Enter,
+    /// The space bar.
+    Space,
+
+    /// The "Compose" key on Linux.
+    Compose,
+
+    Caret,
```

Comment:
what button is caret?

Reply:
Keycodeas are exactly like winit keycode. It should be `^` (shift + 6 on UK keyboard), but it looks like winit doesn't recognize this key via that combination